helper/helper-common/src/main/java/com/linkedin/avroutil1/compatibility/AvroSchemaUtil.java

```diff
@@ -117,6 +174,159 @@ public static Schema findNonNullUnionBranch(Schema schema) {
+      //1st attempt is strict
+      for (Schema readerUnionMember : reader.getTypes()) {
+        SchemaResolutionResult directMatch = resolveReaderVsWriter(writer, readerUnionMember, false, false);
+        if (directMatch != null) {
+          return new SchemaResolutionResult(readerUnionMember, directMatch.getWriterMatch(), false);
+        }
+      }
+      //2nd attempt allowing promotions and aliases
+      for (Schema readerUnionMember : reader.getTypes()) {
+        SchemaResolutionResult looseMatch = resolveReaderVsWriter(writer, readerUnionMember, true, true);
+        //TODO - assert single match?
+        if (looseMatch != null) {
+          return new SchemaResolutionResult(readerUnionMember, looseMatch.getWriterMatch(), false);
+        }
+      }
+    }
+
+    //no unions
+
+    switch (writerType) {
+      //primitives that are promotable (others are handled by default branch):
+      case INT:
+        if (readerType.equals(writerType)) {
+          return new SchemaResolutionResult(reader, writer, false);
+        } else if (usePromotions && INT_PROMOTIONS.contains(readerType)) {
+          return new SchemaResolutionResult(reader, writer, true);
+        }
+        break;
+      case LONG:
+        if (readerType.equals(writerType)) {
+          return new SchemaResolutionResult(reader, writer, false);
+        } else if (usePromotions && LONG_PROMOTIONS.contains(readerType)) {
+          return new SchemaResolutionResult(reader, writer, true);
+        }
+        break;
+      case FLOAT:
+        if (readerType.equals(writerType)) {
+          return new SchemaResolutionResult(reader, writer, false);
+        } else if (usePromotions && readerType == Schema.Type.DOUBLE) {
+          return new SchemaResolutionResult(reader, writer, true);
+        }
+        break;
+      case STRING:
+        if (readerType.equals(writerType)) {
+          return new SchemaResolutionResult(reader, writer, false);
+        } else if (usePromotions && readerType == Schema.Type.BYTES) {
+          return new SchemaResolutionResult(reader, writer, true);
+        }
+        break;
+      case BYTES:
+        if (readerType.equals(writerType)) {
+          return new SchemaResolutionResult(reader, writer, false);
+        } else if (usePromotions && readerType == Schema.Type.STRING) {
+          return new SchemaResolutionResult(reader, writer, true);
+        }
+        break;
+      //named types:
+      case ENUM:
+      case FIXED:
+      case RECORD:
+        if (readerType.equals(writerType)) {
+          //fixed types need to match on size
+          if (writerType == Schema.Type.FIXED) {
+            if (writer.getFixedSize() != reader.getFixedSize()) {
+              return null;
+            }
+          }
+          //named types need to match on fullname, or possibly aliases
+          if (writer.getFullName().equals(reader.getFullName())) {
```

Comment:
In the case of a record or an enum, do we need to check if the fields/values match?

Reply:
not up-front. matching the fields will happen recursively in nested calls. if the fields "match" (taking resolution rules, aliasing and promotions into account) the method will complete, otherwise an exception will be thrown out of nested calls